### PR TITLE
Skip FP registers processing if method don't have floating points use

### LINKED ISSUE #83109: Segregate the register iterations based on intType/floatType
We noticed in https://github.com/dotnet/runtime/pull/79544#issuecomment-1398991764 and lately in #82731 that we should not iterate over all the register sets (specially do not iterate over float/SIMD if we know that there are no variables in the method). This will improve the TP performance.

Today, at multiple hot places, we iterate over all the registers.
```c++
for (regNumber reg = REG_FIRST; reg < ACTUAL_REG_COUNT; reg = REG_NEXT(reg))
{
 ...
}

```

Some other TP improvement potential:
- [x] https://github.com/dotnet/runtime/pull/82731#discussion_r1129794470 (Completed in #83134)
- [ ] https://github.com/dotnet/runtime/pull/82731#discussion_r1129804682

### PR DESCRIPTION
**Pull request by kunalspathak**
Fixes: https://github.com/dotnet/runtime/issues/83109
Contributes to https://github.com/dotnet/runtime/issues/83946

### CODE REVIEW

File: src/coreclr/jit/lsra.h [hunk 1]
```diff
@@ -733,6 +733,7 @@ class LinearScan : public LinearScanInterface
     unsigned int currentSpill[TYP_COUNT];
     bool         needFloatTmpForFPCall;
     bool         needDoubleTmpForFPCall;
+    bool         needFloatingRegisters;
```
Comment (kunalspathak) [inline src/coreclr/jit/lsra.h:736]: A safe change would be to have `needFloatingRegisters |= (theRegisterType != IntRegisterType);` in `newInterval()` method but want to see if my current changes achieve the purpose.

**Comment by kunalspathak** [comment]
Nice [TP Diffs](https://dev.azure.com/dnceng-public/public/_build/results?buildId=261750&view=ms.vss-build-web.run-extensions-tab)

image

File: src/coreclr/jit/lsra.cpp [hunk 2]
```diff
@@ -418,6 +418,8 @@ void LinearScan::updateRegsFreeBusyState(RefPosition& refPosition,
 
 regMaskTP LinearScan::internalFloatRegCandidates()
 {
+    needFloatingRegisters = true;
```
Comment (tannergooding) [inline src/coreclr/jit/lsra.cpp:421]: Does this also cover `maskRegisters`? Should there be a separate concept or should they be tied together?

Comment (kunalspathak) [reply src/coreclr/jit/lsra.cpp:421]: I think they should better be tied together for now. We can revisit it in future.

Comment (tannergooding) [reply src/coreclr/jit/lsra.cpp:421]: Should we maybe call this `needNonGeneralRegisters` or `needNonIntegerRegisters` to possibly be more inclusive and less confusing as to what exactly it covers?

**Comment by tannergooding** [comment]
> Nice [TP Diffs](https://dev.azure.com/dnceng-public/public/_build/results?buildId=261750&view=ms.vss-build-web.run-extensions-tab)

Do we know what causes the potential hit to Arm64 minopts? Seems odd that we see such huge improvements elsewhere, but not always in minopts but only on Arm64

File: src/coreclr/jit/lsra.cpp (hunk 2 above)
Comment (kunalspathak) [reply src/coreclr/jit/lsra.cpp:421]: Sure, will change it.

**Comment by kunalspathak** [comment]
[Jitstress](https://dev.azure.com/dnceng-public/public/_build/results?buildId=261813&view=results) is green.

**Comment by kunalspathak** [comment]
> I see `tests` in particular are the "worst offender", is that maybe due to the number of `HWIntrinsic` tests? If so, do we know why we don't see similar in full-opts?

I will try to find it out.

**Comment by kunalspathak** [comment]
@dotnet/jit-contrib

**Comment by kunalspathak** [comment]
> > I see `tests` in particular are the "worst offender", is that maybe due to the number of `HWIntrinsic` tests? If so, do we know why we don't see similar in full-opts?
> 
> I will try to find it out.

```
PS E:\spmi\pintools\1.0\windows> .\analyze-pin-trace-diff.ps1 -b E:\features\lsra\floatregs\base_arm64_coreclr.txt E:\features\lsra\floatregs\diff_arm64_coreclr.txt
Base: 431066232654, Diff: 431355053761, +0.0670%

?buildIntervals@LinearScan@@QEAAXXZ                                                                                           : 571725277  : +46.55%  : 27.20% : +0.1326%
?processBlockStartLocations@LinearScan@@AEAAXPEAUBasicBlock@@@Z                                                               : 349854912  : +27.75%  : 16.64% : +0.0812%
?resolveRegisters@LinearScan@@QEAAXXZ                                                                                         : 270767431  : +3.38%   : 12.88% : +0.0628%
?impImportAndPushBox@Compiler@@IEAAXPEAUCORINFO_RESOLVED_TOKEN@@@Z                                                            : 2230836    : +6.41%   : 0.11%  : +0.0005%
?impPushOnStack@Compiler@@IEAAXPEAUGenTree@@VtypeInfo@@@Z                                                                     : -2256082   : -0.75%   : 0.11%  : -0.0005%
?impImportCall@Compiler@@IEAA?AW4var_types@@W4opcode_t@@PEAUCORINFO_RESOLVED_TOKEN@@1PEAUGenTree@@HPEAUCORINFO_CALL_INFO@@I@Z : -2438804   : -0.20%   : 0.12%  : -0.0006%
??$allocateRegisters@$0A@@LinearScan@@QEAAXXZ                                                                                 : -436643544 : -1.57%   : 20.77% : -0.1013%
?buildPhysRegRecords@LinearScan@@AEAAXXZ                                                                                      : -464797849 : -100.00% : 22.11% : -0.1078%
```

I investigated the top methods and one thing that stands out is that earlier the `for-loop` of registers was unrolled and now it no longer does, but iterating it fewer times is still better.
image

The call to `buildPhysRegsRecord()` is now getting inlined inside `buildIntervals()`.

image


Other then that, the only extra thing I am doing is setting `needNonIntegerRegisters = true` inside `internalFloatRegCandidates()`.

**Comment by kunalspathak** [comment]
It also occurred to me that for xarch, we already updated the definition of `AVAILABLE_REG_COUNT` to a getter, but for arm64, it was a constant. This PR makes arm64 definition to getter as well and that could reflect the regression on arm64 and not on x64.

File: src/coreclr/jit/lsrabuild.cpp [hunk 3]
```diff
@@ -1940,6 +1940,14 @@ void LinearScan::buildPhysRegRecords()
...
         curr->regOrder  = (unsigned char)i;
     }
+
+    // TODO-CQ: We build physRegRecords before building intervals
+    // and refpositions. During building intervals/refposition, we
+    // would know if there are floating points used. If we can know
+    // that information before we build intervals, we can skip
+    // initializing the floating registers.
+    // For that `compFloatingPointUsed` should be set accurately
+    // before invoking allocator.
```
[the remark this answers is not in the document]
Comment (tannergooding) [reply src/coreclr/jit/lsrabuild.cpp:1950]: Would it instead be possible to have `regMaskTP LinearScan::internalFloatRegCandidates()` lazily do this?

That is, have it do something like:
```cpp
regMaskTP LinearScan::internalFloatRegCandidates()
{
    if (!needNonIntegerRegisters)
    {
        // lazily initialize fpreg state here

        for (unsigned int i = 0; i < lsraRegOrderFltSize; i++)
        {
            regNumber  reg  = lsraRegOrderFlt[i];
            RegRecord* curr = &physRegs[reg];
            curr->regOrder  = (unsigned char)i;
        }

        needNonIntegerRegisters = true;
    }

    if (compiler->compFloatingPointUsed)
    {
        return availableFloatRegs;
    }
    else
    {
        return RBM_FLT_CALLEE_TRASH;
    }
}
```

[the remark this answers is not in the document]
Comment (kunalspathak) [reply src/coreclr/jit/lsrabuild.cpp:1950]: Unfortunately we need the physRegRecord early right after identifying candidates.

[the remark this answers is not in the document]
Comment (kunalspathak) [reply src/coreclr/jit/lsrabuild.cpp:1950]: > But that can't really account for dead code elimination or other factors that might later remove or transforms the nodes to no longer need it.

True, but this is the best we can do I suppose, unless you are proposing to just start from scratch in LSRA and set it to `true` during `BuildNode()` if `tree->GetType() != TYP_INT`.

**Review by tannergooding — APPROVED** [review]

**Review by BruceForstall — APPROVED** [review]

**Review by BruceForstall — APPROVED** [review]

**Pull request merged**